### PR TITLE
 Add support for custom tile size to Zoomify source

### PR DESCRIPTION
**Pull request by lasselaakkonen**
https://github.com/openlayers/openlayers/issues/6608

Thank you for your interest in making OpenLayers better!

In order to get your proposed changes merged into the master branch, we ask you to please make sure the following boxes are checked *before* submitting your pull request.

- [x] This pull request addresses an issue that has been marked with the 'Pull request accepted' label & I have added the link to that issue.
- [x] It contains one or more small, incremental, logically separate commits, with no merge commits.
- [x] I have used clear commit messages.
- [x] Existing tests pass for me locally & I have added or updated tests for new or changed functionality.
- [x] The work herein is covered by a valid [Contributor License Agreement (CLA)](https://github.com/openlayers/cla).

### CODE REVIEW

File: externs/olx.js [hunk 1]
```diff
@@ -7327,6 +7327,15 @@ olx.source.ZoomifyOptions.prototype.transition;
 
 
 /**
+ * Tile size. Same tile size is used for all zoom levels. Default value is
+ * `OpenLayers.DEFAULT_TILE_SIZE`.
+ * @type {number|undefined}
+ * @api
+ */
+olx.source.ZoomifyOptions.prototype.tileSize;
```
Comment (ahocevar) [inline externs/olx.js]: `tileSize` also needs to be added to `olx.source.ZoomifyOptions;` above.

File: src/ol/source/tile.js [hunk 2]
```diff
@@ -75,7 +75,10 @@ ol.source.Tile = function(options) {
    * @protected
    * @type {olx.TileOptions}
    */
-  this.tileOptions = {transition: options.transition};
+  this.tileOptions = {
+    transition: options.transition,
+    tileGrid: this.tileGrid
```
Comment (ahocevar) [inline src/ol/source/tile.js]: This should not be needed.

File: src/ol/imagetile.js [hunk 3]
```diff
@@ -51,6 +52,14 @@ ol.ImageTile = function(tileCoord, state, src, crossOrigin, tileLoadFunction, op
...
   this.tileLoadFunction_ = tileLoadFunction;
 
+  var tileCoordZ = tileCoord[0];
+  var tileGridTileSize = (options.tileGrid ? options.tileGrid.getTileSize(tileCoordZ) : undefined);
+
+  /**
+   * @protected
+   * @type {number}
+   */
+  this.tileSize_ = tileGridTileSize || ol.DEFAULT_TILE_SIZE;
```
Comment (ahocevar) [inline src/ol/imagetile.js]: Why is this needed?

Comment (lasselaakkonen) [reply src/ol/imagetile.js]: I was not completely sure about this change, but right now it is needed because https://github.com/lasselaakkonen/openlayers/blob/b8589acc6e01092cbf85048e350da25c1562c42e/src/ol/source/zoomify.js#L182-L201 needs the value of the configured tile size (previous version always used `ol.DEFAULT_TILE_SIZE`). The `ImageTile` property `tileSize_` does not need to be added if `ol.source.Zoomify.Tile_` can access the `TileGrid` (to get the tile size) and it knows the z-coordinate (for the tile size).

`tileGrid` instead of just `tileSize` is passed to `ImageTile`, so that the tiles can get the tile size based on the z-coordinate. I interpreted that, at least in theory, `TileGrid` can support different tile sizes for different zoom levels.

Comment (ahocevar) [reply src/ol/imagetile.js]: You can change `ol.source.Zoomify` to create a `tileClass` in the `ol.source.Zoomify` constructor,, instead of using `ol.source.Zoomify.Tile_`.

Something like
```diff
diff --git a/src/ol/source/zoomify.js b/src/ol/source/zoomify.js
index 467f8ae..d5e55e1 100644
--- a/src/ol/source/zoomify.js
+++ b/src/ol/source/zoomify.js
@@ -130,6 +130,63 @@ ol.source.Zoomify = function(opt_options) {
 
   var tileUrlFunction = ol.TileUrlFunction.createFromTileUrlFunctions(urls.map(createFromTemplate));
 
+  /**
+   * @constructor
+   * @extends {ol.ImageTile}
+   * @param {ol.TileCoord} tileCoord Tile coordinate.
+   * @param {ol.TileState} state State.
+   * @param {string} src Image source URI.
+   * @param {?string} crossOrigin Cross origin.
+   * @param {ol.TileLoadFunctionType} tileLoadFunction Tile load function.
+   * @param {olx.TileOptions=} opt_options Tile options.
+   * @private
+   */
+  var Tile = function(
+      tileCoord, state, src, crossOrigin, tileLoadFunction, opt_options) {
+
+    ol.ImageTile.call(this, tileCoord, state, src, crossOrigin, tileLoadFunction, opt_options);
+
+    /**
+     * @private
+     * @type {HTMLCanvasElement|HTMLImageElement|HTMLVideoElement}
+     */
+    this.zoomifyImage_ = null;
+
+    /**
+     * @private
+     * @type {ol.Size}
+     */
+    this.tileSize_ = tileGrid ? tileGrid.getTileSize(tileCoord[0]) : ol.DEFAULT_TILE_SIZE;
+
+  };
+  ol.inherits(Tile, ol.ImageTile);
+
+
+  /**
+   * @inheritDoc
+   */
+  Tile.prototype.getImage = function() {
+    if (this.zoomifyImage_) {
+      return this.zoomifyImage_;
+    }
+    var tileSize = this.tileSize_;
+    var image = ol.ImageTile.prototype.getImage.call(this);
+    if (this.state == ol.TileState.LOADED) {
+      if (image.width == tileSize && image.height == tileSize) {
+        this.zoomifyImage_ = image;
+        return image;
+      } else {
+        var context = ol.dom.createCanvasContext2D(tileSize, tileSize);
+        context.drawImage(image, 0, 0);
+        this.zoomifyImage_ = context.canvas;
+        return context.canvas;
+      }
+    } else {
+      return image;
+    }
+  };
+
+
   ol.source.TileImage.call(this, {
     attributions: options.attributions,
     cacheSize: options.cacheSize,
@@ -148,57 +205,6 @@ ol.inherits(ol.source.Zoomify, ol.source.TileImage);
 
 
 /**
- * @constructor
- * @extends {ol.ImageTile}
- * @param {ol.TileCoord} tileCoord Tile coordinate.
- * @param {ol.TileState} state State.
- * @param {string} src Image source URI.
- * @param {?string} crossOrigin Cross origin.
- * @param {ol.TileLoadFunctionType} tileLoadFunction Tile load function.
- * @param {olx.TileOptions=} opt_options Tile options.
- * @private
- */
-ol.source.Zoomify.Tile_ = function(
-    tileCoord, state, src, crossOrigin, tileLoadFunction, opt_options) {
-
-  ol.ImageTile.call(this, tileCoord, state, src, crossOrigin, tileLoadFunction, opt_options);
-
-  /**
-   * @private
-   * @type {HTMLCanvasElement|HTMLImageElement|HTMLVideoElement}
-   */
-  this.zoomifyImage_ = null;
-
-};
-ol.inherits(ol.source.Zoomify.Tile_, ol.ImageTile);
-
-
-/**
- * @inheritDoc
- */
-ol.source.Zoomify.Tile_.prototype.getImage = function() {
-  if (this.zoomifyImage_) {
-    return this.zoomifyImage_;
-  }
-  var tileSize = ol.DEFAULT_TILE_SIZE;
-  var image = ol.ImageTile.prototype.getImage.call(this);
-  if (this.state == ol.TileState.LOADED) {
-    if (image.width == tileSize && image.height == tileSize) {
-      this.zoomifyImage_ = image;
-      return image;
-    } else {
-      var context = ol.dom.createCanvasContext2D(tileSize, tileSize);
-      context.drawImage(image, 0, 0);
-      this.zoomifyImage_ = context.canvas;
-      return context.canvas;
-    }
-  } else {
-    return image;
-  }
-};
-
-
-/**
  * @enum {string}
  * @private
  */


```

File: externs/olx.js (hunk 1 above)
Comment (lasselaakkonen) [reply externs/olx.js]: I have now fixed this.

File: src/ol/imagetile.js (hunk 3 above)
Comment (lasselaakkonen) [reply src/ol/imagetile.js]: Exactly.

I have now removed changes from all files other than Zoomify source.

When I put `ol.source.Zoomify.Tile_` constructor inside `ol.source.Zoomify` constructor, checking/compiling kept on throwing errors at me:

```
ERR! compile /var/folders/bs/ry1bvzld073_mj0s02nt_p740000gn/T/exports117931-23665-1ow79as.bp33i.js:9181: ERROR - Parse error. ',' expected
ERR! compile     ol.source.Zoomify~ZoomifyTileClass.prototype,
ERR! compile                      ^
```

So now `ZoomifyTileClass` is still defined outside `ol.source.Zoomify` and now `TileGrid` is bound to it.

Comment (ahocevar) [reply src/ol/imagetile.js]: In this case, you should leave the prototype as `ol.source.Zoomify.Tile_` and bind to a local variable in the `ol.source.Zoomify` constructor. Basically exactly the other way around than what you currently have. Otherwise `ol.source.Zoomify.Tile_` will only work as expected when there is only one Zoomify source in your app.

**Comment by lasselaakkonen** [comment]
@ahocevar Thank you for reviewing this pull request!

**Comment by ahocevar** [comment]
Thanks @lasselaakkonen for this excellent first-time contribution.
